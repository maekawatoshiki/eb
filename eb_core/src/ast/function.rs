use crate::lexer::location::Location;

#[derive(Debug, Clone, PartialEq)]
pub struct Node {
    name: String,
    loc: Location,
}

impl Node {
    pub fn new(name: String, loc: Location) -> Self {
        Self { name, loc }
    }
}

impl From<Node> for super::Node {
    fn from(node: Node) -> Self {
        super::Node::Function(node)
    }
}
