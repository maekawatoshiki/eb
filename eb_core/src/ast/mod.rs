pub mod expr;
pub mod function;

#[derive(Debug, PartialEq, Clone)]
pub enum Node {
    Function(function::Node),
    Expr(expr::Node),
}
