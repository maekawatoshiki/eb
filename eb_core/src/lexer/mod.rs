pub mod location;
pub mod source;
pub mod token;

use source::Source;

pub fn tokenize<'a>(source: &'a Source) -> token::TokenStream<'a> {
    token::TokenStream::new(source)
}

#[test]
fn tokenize1() {
    use location::Location;
    use token::Token;

    let source = Source::String(r#"func f(x i32) i32: x;;"#.to_string());
    let tokenize: Vec<Token> = tokenize(&source).into_iter().collect();
    let correct = vec![
        Token::new(token::TokenKind::Ident("func"), Location(0)),
        Token::new(token::TokenKind::Ident("f"), Location(4)),
        Token::new(
            token::TokenKind::OpenDelim(token::DelimKind::Paren),
            Location(6),
        ),
        Token::new(token::TokenKind::Ident("x"), Location(7)),
        Token::new(token::TokenKind::Ident("i32"), Location(8)),
        Token::new(
            token::TokenKind::CloseDelim(token::DelimKind::Paren),
            Location(12),
        ),
        Token::new(token::TokenKind::Ident("i32"), Location(13)),
        Token::new(token::TokenKind::Colon, Location(17)),
        Token::new(token::TokenKind::Ident("x"), Location(18)),
        Token::new(token::TokenKind::DoubleSemicolon, Location(20)),
    ];
    assert_eq!(tokenize.len(), correct.len());
    assert!(tokenize.iter().zip(correct.iter()).all(|(a, b)| a == b))
}

#[test]
fn tokenize2() {
    use location::Location;
    use token::{BinOpKind, DelimKind, Token, TokenKind};

    let source = Source::String(
        r#"
        func f(x y i32) i32: 
            x + 1 == y - 1;;
        "#
        .to_string(),
    );
    let tokenize: Vec<Token> = tokenize(&source).into_iter().collect();
    for token in &tokenize {
        println!("{:?}", token);
    }
    let correct = vec![
        Token::new(TokenKind::Ident("func"), Location(0)),
        Token::new(TokenKind::Ident("f"), Location(13)),
        Token::new(TokenKind::OpenDelim(DelimKind::Paren), Location(15)),
        Token::new(TokenKind::Ident("x"), Location(16)),
        Token::new(TokenKind::Ident("y"), Location(17)),
        Token::new(TokenKind::Ident("i32"), Location(19)),
        Token::new(TokenKind::CloseDelim(DelimKind::Paren), Location(23)),
        Token::new(TokenKind::Ident("i32"), Location(24)),
        Token::new(TokenKind::Colon, Location(28)),
        Token::new(TokenKind::Ident("x"), Location(29)),
        Token::new(TokenKind::BinOp(BinOpKind::Plus), Location(44)),
        Token::new(TokenKind::Int("1"), Location(46)),
        Token::new(TokenKind::BinOp(BinOpKind::Eq), Location(48)),
        Token::new(TokenKind::Ident("y"), Location(51)),
        Token::new(TokenKind::BinOp(BinOpKind::Minus), Location(53)),
        Token::new(TokenKind::Int("1"), Location(55)),
        Token::new(TokenKind::DoubleSemicolon, Location(57)),
    ];
    assert_eq!(tokenize.len(), correct.len());
    assert!(tokenize.iter().zip(correct.iter()).all(|(a, b)| a == b))
}
