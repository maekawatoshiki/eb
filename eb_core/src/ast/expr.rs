use crate::lexer::{location::Location, token::PunctKind};

#[derive(Debug, Clone, PartialEq)]
pub struct Node {
    kind: Kind,
    loc: Location,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Kind {
    Int(i64),
    Ident(String),
    BinOp(BinOpKind, Box<Node>, Box<Node>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Eq,
}

impl Node {
    pub fn new(kind: Kind, loc: Location) -> Self {
        Self { kind, loc }
    }
}

impl From<Node> for super::Node {
    fn from(node: Node) -> Self {
        super::Node::Expr(node)
    }
}

impl From<PunctKind> for BinOpKind {
    fn from(p: PunctKind) -> Self {
        match p {
            PunctKind::Plus => Self::Add,
            PunctKind::Minus => Self::Sub,
            PunctKind::Eq => Self::Eq,
            _ => panic!(),
        }
    }
}
