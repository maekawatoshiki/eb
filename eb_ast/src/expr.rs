use super::function;
use lexer::location::Location;

#[derive(Debug, Clone, PartialEq)]
pub struct Node {
    kind: Kind,
    loc: Location,
}

#[derive(Debug, Clone, PartialEq)]
pub enum Kind {
    Int(i64),
    Ident(String),
    Function(Box<function::Node>),
    BinOp(BinOpKind, Box<Node>, Box<Node>),
    Call(Box<Node>, Vec<Node>),
    If(Box<Node>, Box<Node>, Option<Box<Node>>),
    Exprs(Vec<Node>),
}

#[derive(Debug, Clone, PartialEq)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Eq,
    Neq,
}

impl Node {
    pub fn new(kind: Kind, loc: Location) -> Self {
        Self { kind, loc }
    }
}
