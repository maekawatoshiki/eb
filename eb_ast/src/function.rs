use super::expr;

#[derive(Debug, Clone, PartialEq)]
pub struct Node {
    name: String,
    params: Vec<Param>,
    body: Vec<expr::Node>,
}

#[derive(Debug, Clone, PartialEq)]
pub struct Param {
    name: String,
}

impl Node {
    pub fn new(name: String, params: Vec<Param>, body: Vec<expr::Node>) -> Self {
        Self { name, params, body }
    }
}

impl Param {
    pub fn new(name: String) -> Self {
        Self { name }
    }
}
