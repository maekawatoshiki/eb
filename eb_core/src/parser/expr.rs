use super::Context;
use crate::ast::expr;
use anyhow::Result;

pub fn parse(ctx: &mut Context) -> Result<expr::Node> {
    let loc = ctx.cur_loc()?;
    let ident = ctx
        .expect_any_ident()?
        .kind()
        .as_ident()
        .unwrap()
        .to_string();
    Ok(expr::Node::new(expr::Kind::Ident(ident), loc))
}
