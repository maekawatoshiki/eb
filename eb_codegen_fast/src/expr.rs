use super::function::{self, Context};
use anyhow::Result;
use ast::expr as ast_expr;
use vm::inst::Inst;

pub fn visit(ctx: &mut Context, expr: &ast_expr::Node) -> Result<()> {
    match expr.kind() {
        ast_expr::Kind::Int(i) => {
            ctx.push(Inst::PushInt(*i as i32)); // TODO
        }
        ast_expr::Kind::Ident(ident) => {
            ctx.push(Inst::PushStr(ident.to_owned()));
        }
        ast_expr::Kind::BinOp(op, lhs, rhs) => visit_binop(ctx, op, lhs, rhs)?,
        ast_expr::Kind::Function(func) => {
            let mut ctx_ = Context::default();
            function::visit(&mut ctx_, func)?;
            ctx.add_child(ctx_);
        }
        ast_expr::Kind::Call(callee, args) => visit_call(ctx, callee, args)?,
        ast_expr::Kind::If(cond, then_, else_) => visit_if(ctx, cond, then_, else_)?,
        ast_expr::Kind::Return(val) => visit_ret(ctx, val)?,
        ast_expr::Kind::Exprs(exprs) => {
            for expr in exprs {
                visit(ctx, expr)?;
            }
        }
    }
    Ok(())
}

fn visit_binop(
    ctx: &mut Context,
    op: &ast_expr::BinOpKind,
    lhs: &ast_expr::Node,
    rhs: &ast_expr::Node,
) -> Result<()> {
    visit(ctx, lhs)?;
    visit(ctx, rhs)?;
    match op {
        ast_expr::BinOpKind::Sub => ctx.push(Inst::Sub),
        ast_expr::BinOpKind::Mul => ctx.push(Inst::Mul),
        ast_expr::BinOpKind::Eq => ctx.push(Inst::Eq),
        _ => todo!(),
    }
    Ok(())
}

fn visit_call(ctx: &mut Context, callee: &ast_expr::Node, args: &[ast_expr::Node]) -> Result<()> {
    for arg in args {
        visit(ctx, arg)?;
    }
    match callee.kind() {
        ast_expr::Kind::Ident(name) => {
            ctx.push(Inst::PushStr(name.to_owned()));
        }
        _ => todo!(),
    }
    ctx.push(Inst::Call);
    Ok(())
}

fn visit_if(
    ctx: &mut Context,
    cond: &ast_expr::Node,
    then_: &ast_expr::Node,
    else_: &Option<Box<ast_expr::Node>>,
) -> Result<()> {
    assert!(else_.is_none());
    visit(ctx, cond)?;
    let cur = ctx.code.len() as i32;
    ctx.push(Inst::Jne(0));
    visit(ctx, then_)?;
    let merge = ctx.code.len() as i32;
    match ctx.code.get_mut(cur as usize).unwrap() {
        Inst::Jne(ref mut offset) => {
            *offset = merge - cur;
        }
        _ => panic!(),
    }
    Ok(())
}

fn visit_ret(ctx: &mut Context, val: &ast_expr::Node) -> Result<()> {
    visit(ctx, val)?;
    ctx.push(Inst::Ret);
    Ok(())
}
