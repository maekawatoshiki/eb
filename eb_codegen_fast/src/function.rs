use super::expr;
use anyhow::Result;
use ast::function as func;
use vm::inst::{Code, Inst};

#[derive(Debug)]
pub struct Context {
    pub name: String,
    pub param_names: Vec<String>,
    pub code: Code,
    pub children: Vec<Context>,
}

pub fn visit(ctx: &mut Context, func: &func::Node) -> Result<()> {
    ctx.name = func.name().to_owned();
    ctx.param_names = func.params().iter().map(|p| p.name().to_owned()).collect();
    for expr in func.body() {
        expr::visit(ctx, expr)?;
    }
    Ok(())
}

impl Default for Context {
    fn default() -> Self {
        Self {
            name: "".to_owned(),
            param_names: vec![],
            code: Code(vec![]),
            children: vec![],
        }
    }
}

impl Context {
    pub fn push(&mut self, inst: Inst) {
        self.code.0.push(inst)
    }

    pub fn add_child(&mut self, ctx: Self) {
        self.children.push(ctx)
    }
}

#[cfg(test)]
mod test {
    extern crate eb_lexer as lexer;
    extern crate eb_parser as parser;
    extern crate insta;
    use super::*;
    use lexer::{source::Source, tokenize};
    use parser::{function::parse, Context as ParserContext};

    #[test]
    fn codegen1() {
        let source = Source::String(r#"func f(): ;;"#.to_string());
        let mut ctx = ParserContext::new(tokenize(&source));
        let node = parse(&mut ctx).expect("fail to parse");
        let mut ctx = Context::default();
        visit(&mut ctx, &node).unwrap();
        insta::assert_debug_snapshot!(ctx);
    }

    #[test]
    fn codegen2() {
        let source = Source::String(r#"func f(x): x;;"#.to_string());
        let mut ctx = ParserContext::new(tokenize(&source));
        let node = parse(&mut ctx).expect("fail to parse");
        let mut ctx = Context::default();
        visit(&mut ctx, &node).unwrap();
        insta::assert_debug_snapshot!(ctx);
    }

    #[test]
    fn codegen3() {
        let source = Source::String(
            r#"
            func f(x): 
                if x == 1:
                    return 1 ;;
                x * f(x - 1) ;;"#
                .to_string(),
        );
        let mut ctx = ParserContext::new(tokenize(&source));
        let node = parse(&mut ctx).expect("fail to parse");
        let mut ctx = Context::default();
        visit(&mut ctx, &node).unwrap();
        insta::assert_debug_snapshot!(ctx);
    }
}
