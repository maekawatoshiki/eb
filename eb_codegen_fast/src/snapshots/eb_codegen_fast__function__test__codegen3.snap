---
source: src/function.rs
expression: ctx

---
FunctionContext {
    name: "f",
    param_names: [
        "x",
    ],
    code: Code(
        [
            Get(
                "x",
            ),
            PushInt(
                1,
            ),
            Eq,
            Jne(
                3,
            ),
            PushInt(
                1,
            ),
            Ret,
            Get(
                "x",
            ),
            Get(
                "x",
            ),
            PushInt(
                1,
            ),
            Sub,
            Get(
                "f",
            ),
            Call,
            Mul,
        ],
    ),
    children: [],
}
